src/passes/Unsubtyping.cpp

```diff
@@ -193,6 +193,22 @@ struct Unsubtyping
     noteSubtype(sub->type, super->type);
   }
 
+  void noteNonFlowSubtype(Expression* sub, Type super) {
```

Comment:
IIRC, nothing about Unsubtyping is flow-sensitive in any way, so differentiating between non-flow subtyping and flow subtyping seems like the wrong abstraction, or rather a fix for a problem we do not have here.

Is it the case that `noteSubtype` was simply never passed a basic supertype before that change to `SubtypeDiscoverer`? If so, I suspect the proper fix is to change `isBottom()` to `isBasic()` in the check at the beginning of `noteSubtype`.

Reply:
Heh, our thinking is identical: I actually considered just that, but sadly it doesn't work: we *do* need to consider basic types there. Consider if we write a user type to `anyref`, then that interacts with casts. The cast logic basically needs to know about all the things that might flow into `anyref`, since it needs to preserve cast differences among them, when we cast back from `anyref`. That is, unfortunately, where the flow sensitivity comes from.

(And for that reason the new test in this PR must contain casts.)

Reply:
(friendly ping @tlively , there is at least one project that regressed by this that is eager to get a fix)

Reply:
Oops, sorry for letting this slip.